### PR TITLE
Made nested unroll work.

### PR DESCRIPTION
**Pull request by adam-yang**

### CODE REVIEW

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp [hunk 1]
```diff
@@ -673,6 +680,44 @@ bool DxilLoopUnroll::IsLoopSafeToClone(Loop *L) {
   return true;
 }
 
+void DxilLoopUnroll::RecursivelyCreateSubLoopForIteration(LPPassManager &LPM, LoopInfo *LI, Loop *OuterL, Loop *L, LoopIteration &Iter) {
```
Comment (pow2clk) [inline lib/Transforms/Scalar/DxilLoopUnroll.cpp:683]: As I read this, it replaces the loop that previously contained the subloop `L` with a new dummy loop that just contains `L` and no other subloops. Is that correct? I don't think the name of the function is particularly helpful to understanding that. Calling it something like `ReplaceParentWithDummy` or adding a comment might clarify things

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp [hunk 2]
```diff
@@ -673,6 +680,44 @@ bool DxilLoopUnroll::IsLoopSafeToClone(Loop *L) {
...
+  }
+  else {
+    LI->addTopLevelLoop(NewL);
+  }
+
+  for (auto it = L->block_begin(), end = L->block_end(); it != end; it++) {
+    BasicBlock *OriginalBB = *it;
+    BasicBlock *NewBB = cast<BasicBlock>(Iter.VarMap[OriginalBB]);
+    // addBlockEntry instead of addBasicBlockToLoop because addBasicBlockToLoop
+    // will try to add block to outer loop, which adds it twice.
```
Comment (pow2clk) [inline lib/Transforms/Scalar/DxilLoopUnroll.cpp:700]: Doesn't the code below add it to the outer loop too?

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp [hunk 3]
```diff
@@ -1018,7 +1068,13 @@ bool DxilLoopUnroll::runOnLoop(Loop *L, LPPassManager &LPM) {
       for (size_t i = 0; i < Iterations.size(); i++) {
         LoopIteration &Iteration = *Iterations[i].get();
         for (BasicBlock *BB : Iteration.Body) {
-          if (!Iteration.Extended.count(BB)) {
+          if (!Iteration.Extended.count(BB) &&
+            // Check We didn't already add BB to a sub-loop
+            !LI->getLoopFor(BB) &&
+            // Check if it's already added to loop when we added
+            // the block to new child loop earlier.
+            !OuterL->contains(BB))
```
Comment (pow2clk) [inline lib/Transforms/Scalar/DxilLoopUnroll.cpp:1076]: Is there a correct case where it has been added to a subloop, but that subloop isn't part of the outer loop?

What I'm asking is whether it's possible for `LI->getLoopFor(BB)  && !OuterL->contains(BB)` to be true. If not, I think you can just check the `OuterL->contains(BB)` and maybe assert the other part?

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp [hunk 4]
```diff
@@ -633,16 +635,21 @@ static bool BreakUpArrayAllocas(bool AllowOOBIndex, IteratorT ItBegin, IteratorT
...
 
   // Must remove all child loops first.
   for (Loop *SubL : SubLoops) {
-    RecursivelyRemoveLoopFromQueue(LPM, SubL);
+    RecursivelyRemoveLoopOnSuccess(LPM, SubL);
   }
 
+  // Remove any loops/subloops that failed because we are about to
+  // delete the loops, so gotta avoid dangling pointers
+  LoopsThatFailed.erase(L);
```
Comment (pow2clk) [inline lib/Transforms/Scalar/DxilLoopUnroll.cpp:650]: What is the mechanism whereby a loop that failed would get retried?

**Review by pow2clk — COMMENTED** [review]
I understand the big picture here, but I would appreciate a few small clarifications

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp (hunk 4 above)
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilLoopUnroll.cpp:650]: If a loop unroll succeeds, all the blocks and its child loops get deleted. But before that happens new sub Loop structures are created for the cloned blocks (which happens in RecursivelyCreateSubLoopForIteration). Those will get retried.

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp (hunk 3 above)
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilLoopUnroll.cpp:1080]: In an older version of this change, RecursivelyRecreateSubLoopForIteration didn't add all blocks to all outer loops. That's no longer the case.

Comment (adam-yang) [reply lib/Transforms/Scalar/DxilLoopUnroll.cpp:1080]: Fixed.

**Review by tex3d — APPROVED** [review]
Looks good.

**Pull request merged**